q2_diversity_lib/_util.py

```diff
@@ -22,17 +31,79 @@ def _drop_undefined_samples(counts: np.ndarray, sample_ids: np.ndarray,
     return (filtered_counts, filtered_sample_ids)
 
 
-def _disallow_empty_tables(some_function):
+def _disallow_empty_tables_passed_object(some_function):
```

Comment:
Personally I would consolidate `_disallow_empty_tables_passed_object` and `_disallow_empty_tables_passed_filepath` into one decorator function, maybe something like the following (untested code):

```python
@decorator.decorator
def disallow_empty_table(func, *args, **kwargs):
    bound_signature = signature(func).bind(*args, **kwargs)
    table = bound_signature.arguments.get('table')
    if table is None:
        raise TypeError("The wrapped function has no table parameter.")

    if isinstance(table, str) or isinstance(table, BIOMV210Format):
        table = str(table)
        if not os.path.exists(table):
            raise ValueError('%s is not a valid filepath')
        table_obj = load_table(table)
    elif isinstance(table, Table):
        table_obj = table
    else:
        raise ValueError("invalid view type")

    if table_obj.is_empty():
        raise ValueError('The provided table is empty')
    
    return func(*args, **kwargs)
```

This might make your life a bit easier.

Reply:
Thanks for pointing me at `arguments.get()`. Dropping the try/except block cleans this up so much! 

Am I reading this right, that within the decorator, we're coercing any table-like into biom format so that we can use biom's `is_empty()` for our test, but that the decorator itself returns the function and its arguments unmodified?

Reply:
> Am I reading this right, that within the decorator, we're coercing any table-like into biom format so that we can use biom's is_empty() for our test, but that the decorator itself returns the function and its arguments unmodified?

Yes! We might need to rethink that strategy at a later date - not sure how much of a penalty we incur by just discarding, but worth keeping mind if performance becomes problematic.